lib/helper/Testcafe.js

```diff
@@ -0,0 +1,661 @@
+      // .video(global.output_dir)
+      .browsers(this.options.show ? this.options.browser : `${this.options.browser}:headless`)
+      .reporter('minimal')
+      .run({
+        skipJsErrors: true,
+        skipUncaughtErrors: true,
+        quarantineMode: true,
+        selectorTimeout: this.options.waitForTimeout,
+      })
+      .catch((err) => {
+        this.debugSection('_before', `Error ${err.toString()}`);
+        this.isRunning = false;
+        this.testcafe.close();
+      });
+
+    this.t = await testControllerHolder.get();
+    assert(this.t, 'Expected to have the testcafe test controller');
+
+    if (this.options.windowSize && this.options.windowSize.indexOf('x') > 0) {
+      const dimensions = this.options.windowSize.split('x');
+      await this.t.resizeWindow(parseInt(dimensions[0], 10), parseInt(dimensions[1], 10));
+    }
+  }
+
+  async _stopBrowser() {
+    this.debugSection('_after', 'Stopping testcafe browser...');
+
+    testControllerHolder.free();
+    if (this.testcafe) {
+      this.testcafe.close();
+    }
+
+    fs.unlinkSync(this.dummyTestcafeFile); // remove the dummy test
+    this.t = undefined;
+
+    this.isRunning = false;
+  }
+
+  _init() {}
+
+  async _beforeSuite() {
+    if (!this.options.restart && !this.options.manualStart && !this.isRunning) {
+      this.debugSection('Session', 'Starting singleton browser session');
+      return this._startBrowser();
+    }
+  }
+
+
+  async _before() {
+    if (this.options.restart && !this.options.manualStart) return this._startBrowser();
+    if (!this.isRunning && !this.options.manualStart) return this._startBrowser();
+  }
+
+  async _after() {
+    if (!this.isRunning) return;
+
+    if (this.options.restart) {
+      this.isRunning = false;
+      return this._stopBrowser();
+    }
+
+    if (this.options.keepBrowserState) return;
+
+    if (!this.options.keepCookies) {
+      this.debugSection('Session', 'cleaning cookies and localStorage');
+      await this.clearCookie();
+
+      // TODO IMHO that should only happen when
+      await this.executeScript(() => localStorage.clear())
+        .catch((err) => {
+          if (!(err.message.indexOf("Storage is disabled inside 'data:' URLs.") > -1)) throw err;
+        });
+    }
+  }
+
+  _afterSuite() {
+  }
+
+  async _finishTest() {
+    if (!this.options.restart && this.isRunning) return this._stopBrowser();
+  }
+
+  /**
+   * Get elements by different locator types, including strict locator
+   * Should be used in custom helpers:
+   *
+   * ```js
+   * const elements = await this.helpers['Testcafe']._locate({name: 'password'});
+   * ```
+   *
+   * {{ react }}
+   */
+  async _locate(locator) {
+    return findElements(this.context, locator);
+  }
+
+  /**
+   * {{> amOnPage }}
+   */
+  async amOnPage(url) {
+    if (!(/^\w+\:\/\//.test(url))) {
+      url = this.options.url + url;
+    }
+
+    return this.t.navigateTo(url)
+      .catch(mapError);
```

Comment:
Btw, why do we need that `mapError` function? Can we just throw TestCafe exceptions?

Reply:
Testcafe methods throw an error object with lots of information which is quite unreadable if it goes unmapped. However I will see if I can find a better solution.